overlord/snapstate/autorefresh_gating.go

```diff
@@ -32,6 +35,289 @@ import (
+		if !ok {
+			hold = &holdInfo{
+				FirstHeld: now,
+			}
+		}
+
+		maxDur := maxAllowedPostponement(gatingSnap, heldSnap)
+
+		// calculate max hold duration that's left considering previous hold
+		// requests of this snap. This doesn't take last refresh into consideration
+		// yet (that's handled further down), but it makes sure than a cumulative
+		// time the given gating snap holds another snap (or self) is within
+		// maxAllowedPostponement limit.
+		left := maxDur - now.Sub(hold.FirstHeld)
+		if left <= 0 {
+			herr.SnapsInError[heldSnap] = fmt.Errorf("snap %q cannot hold snap %q anymore", gatingSnap, heldSnap)
+			continue
+		}
+
+		dur := holdDuration
+		if dur == 0 {
+			// duration not specified, using a default one (maximum) or what's
+			// left of it.
+			dur = left
+		} else {
+			// explicit hold duration requested
+			if dur > maxDur {
+				herr.SnapsInError[heldSnap] = fmt.Errorf("requested holding duration %s exceeds maximum holding for snap %q", holdDuration, heldSnap)
+				continue
+			}
+		}
+
+		lastRefreshTime, err := lastRefreshed(st, heldSnap)
+		if err != nil {
+			return err
+		}
+
+		newHold := now.Add(dur)
+		// consider last refresh time and adjust hold duration if needed so it's
+		// not exceeded.
+		switch {
+		case newHold.Before(lastRefreshTime.Add(maxPostponement)):
+			hold.HoldUntil = newHold
+		case holdDuration == 0:
+			// default duration requested but it exceeds max
+			// postponement, adjust it.
+			hold.HoldUntil = lastRefreshTime.Add(maxPostponement)
+		default:
+			herr.SnapsInError[heldSnap] = fmt.Errorf("requested holding duration %s exceeds maximum total holding for snap %q", holdDuration, heldSnap)
+			continue
+		}
+
+		// finally store/update gating hold data
+		if _, ok := gating[heldSnap]; !ok {
+			gating[heldSnap] = make(map[string]*holdInfo)
+		}
+		gating[heldSnap][gatingSnap] = hold
+	}
+
+	if len(herr.SnapsInError) != len(affectingSnaps) {
+		st.Set("snaps-hold", gating)
+	}
+	if len(herr.SnapsInError) > 0 {
+		return herr
+	}
+	return nil
+}
+
+// ProceedWithRefresh unblocks all snaps held by gatingSnap for refresh. This
+// should be called for --proceed on the gatingSnap.
+func ProceedWithRefresh(st *state.State, gatingSnap string) error {
+	gating, err := refreshGating(st)
+	if err != nil {
+		return err
+	}
+	if len(gating) == 0 {
+		return nil
+	}
+
+	var changed bool
+	for heldSnap, gatingSnaps := range gating {
+		if _, ok := gatingSnaps[gatingSnap]; ok {
+			delete(gatingSnaps, gatingSnap)
+			changed = true
+		}
+		if len(gating[heldSnap]) == 0 {
+			delete(gating, heldSnap)
+		}
+	}
+
+	if changed {
+		st.Set("snaps-hold", gating)
+	}
+	return nil
+}
+
+// pruneGating resets gating information by:
+// - removing affecting snaps whose held time expired.
+// - removing affecting snaps that are not in candidates (meaning there is no update for them anymore).
+func pruneGating(st *state.State, candidates map[string]*refreshCandidate) error {
```

Comment:
I'm still a bit unsure how this will be used in the end, it's good to have a sketch but maybe we should leave this out until we have a PR using it.

Reply:
Sure, I'm fine to leave it out if that helps, will remove it for now.